src/keywrap/pkcs11/mod.rs

```diff
@@ -0,0 +1,333 @@
+        let json_str = encrypt_multiple(&pkcs11_recipients, opts_data)?;
+
+        Ok(json_str)
+    }
+
+    fn unwrap_keys(&self,
+                   dc: &DecryptConfig,
+                   annotation: &[u8],
+    ) -> Result<Vec<u8>> {
+
+        let mut pkcs11_keys = Vec::new();
+
+        let priv_keys = match self.private_keys(&dc.param) {
+            Some(x) => x,
+            None => return Err(anyhow!("")),
+        };
+
+        let p11conf_opt = p11conf_from_params(&dc.param)?;
+
+        for key in priv_keys {
+
+            let mut k = parse_private_key(&key, &vec![], "PKCS11".to_string())?;
+            match k {
+                KeyType::rpk(r) => {
+                },
+                KeyType::pkfo(mut p) => {
+                    if let Some(ref p11conf) = p11conf_opt {
+                        p.uriw.set_module_directories(&p11conf.module_directories);
+                        p.uriw.set_allowed_module_paths(&p11conf.allowed_module_paths);
+                        pkcs11_keys.push(p);
+                    }
+                },
+            }
+        }
+
+        let plaintext = decrypt_pkcs11(&pkcs11_keys, annotation)?;
+
+        Ok(plaintext)
+    }
+
+
+    fn annotation_id(&self) -> &str {
+        "org.opencontainers.image.enc.keys.pkcs11"
+    }
+
+    fn no_possible_keys(&self,
+                        dcparameters: &HashMap<String, Vec<Vec<u8>>>
+    ) -> bool {
+        self.private_keys(dcparameters).is_none()
+    }
+
+    fn private_keys(&self,
+                    dcparameters: &HashMap<String, Vec<Vec<u8>>>,
+    ) -> Option<Vec<Vec<u8>>> {
+        dcparameters.get("pkcs11-yamls").cloned()
+    }
+
+    fn recipients(&self, _packet: String) -> Option<Vec<String>> {
+        Some(vec!["[pkcs11]".to_string()])
+    }
+}
+
+
+fn p11conf_from_params(dcparameters: &HashMap<String, Vec<Vec<u8>>>,
+) -> Result<Option<Pkcs11Config>> {
+    if dcparameters.contains_key("pkcs11-config") {
+        return Ok(Some(parse_pkcs11_config_file(&dcparameters["pkcs11-config"][0])?))
+    }
+    Ok(None)
+}
+
+fn add_pub_keys(dc: &DecryptConfig,
+                pubkeys: &Vec<Vec<u8>>,
+) -> Result<Vec<KeyType>> {
+    let mut pkcs11_keys = Vec::<KeyType>::new();
+    if pubkeys.is_empty() {
+        return Ok(pkcs11_keys);
+    }
+
+    let p11conf_opt = p11conf_from_params(&dc.param)?;
+
+    for pubkey in pubkeys {
+        let mut k = parse_public_key(pubkey, "PKCS11".to_string())?;
+        match &mut k {
+            KeyType::rpk(r) => {},
+            KeyType::pkfo(p) => {
+                if let Some(ref p11conf) = p11conf_opt {
+                    p.uriw.set_module_directories(&p11conf.module_directories);
+                    p.uriw.set_allowed_module_paths(&p11conf.allowed_module_paths);
+                }
+            },
+        }
+        pkcs11_keys.push(k);
+    }
+
+    Ok(pkcs11_keys)
+}
+
+
+
+
+#[cfg(test)]
+mod kw_tests {
+    use super::*;
+
+    //const SOFTHSM_SETUP: &str = "../../scripts/softhsm_setup";
```

Comment:
Should we commit a copy of the softhsm_setup script? It currently exists in the ocicrypt golang repo.